src/metldata/builtin_transformations/infer_relation/data_transform.py

```diff
@@ -0,0 +1,121 @@
+    MutableClassResources,
+    MutableDatapack,
+    ResourceId,
+)
+from metldata.builtin_transformations.common.utils import data_to_dict
+from metldata.builtin_transformations.infer_relation.config import (
+    InferRelationConfig,
+)
+from metldata.builtin_transformations.infer_relation.resolve_path import (
+    resolve_path,
+)
+from metldata.transform.exceptions import EvitableTransformationError
+
+
+def get_class_resources(
+    *, data: MutableDatapack, class_name: str
+) -> MutableClassResources:
+    """Extract the resources of a given class from the dictionary."""
+    resources = data["resources"].get(class_name)
+    if not resources:
+        raise EvitableTransformationError()
+    return resources
+
+
+def infer_data_relation(
+    *, data: DataPack, model: SchemaPack, transformation_config: InferRelationConfig
+) -> DataPack:
+    """Adds inferred relations to the data as given in the configuration."""
+    modified_data = data_to_dict(data)
+
+    # name of the class to which the relation will be added
+    class_name = transformation_config.class_name
+
+    # name of the relation to be added
+    relation_name = transformation_config.relation_name
+
+    path = transformation_config.relation_path
+    referenced_class = path.target
+
+    # the target prefix refers to resources that will be modified by the transformation
+    target_resources = get_class_resources(data=modified_data, class_name=class_name)
+
+    for resource_id, resource in target_resources.items():
+        relation_target_ids = resolve_path(
+            data=data, source_resource_id=resource_id, path=path
+        )
+        type_corrected_relation_target_ids = modify_relation_target_ids_type(
+            model, class_name, relation_name, relation_target_ids
+        )
+        # add a new relation to that resource
+        resource.setdefault("relations", {})[relation_name] = {
+            "targetClass": referenced_class,
+            "targetResources": type_corrected_relation_target_ids,
+        }
+    return DataPack.model_validate(modified_data)
+
+
+def modify_relation_target_ids_type(
+    model: SchemaPack,
+    class_name: str,
+    relation_name: str,
+    relation_target_ids: frozenset[ResourceId],
+) -> frozenset[ResourceId] | ResourceId | None:
+    """Post transformation modification of the relation target IDs type based on
+    relation specification of the transformed model.
+
+    This function does not handle the case where `relation_target_ids` is expected
+    to contain a single item, but `resolve_path` returns multiple values. Such cases
+    will trigger a validation error in post-transformation validation.
+    """
+    class_definition = get_class_definition(model, class_name)
+
+    relation_spec = get_relation_specification(class_definition, relation_name)
+
+    # If relation expects single target and relation_target_ids has 0 or 1 item,
+    # return single item or None
```

Comment:
Would checking with the spec for `mandatory` and aborting early  if `None` is returned in that case make sense?
I guess that's already handled by the validation at the end either way...

Reply:
No not really. It is already compact and accurate this way. As you say, validations handle it.